// @schibstedspain/rosetta/lib/i18n.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x4, _x5, _x6) { var _again = true; _function: while (_again) { var object = _x4, property = _x5, receiver = _x6; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x4 = parent; _x5 = property; _x6 = receiver; _again = true; desc = parent = undefined; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj['default'] = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _events = require('events');

var _adaptersDefault = require('./adapters/default');

var _adaptersDefault2 = _interopRequireDefault(_adaptersDefault);

var _decorator = require('./decorator');

var decorator = _interopRequireWildcard(_decorator);

var _slugify = require('./slugify');

var _slugify2 = _interopRequireDefault(_slugify);

var CHANGE_TRANSLATION_EVENT = 'translations';

exports.CHANGE_TRANSLATION_EVENT = CHANGE_TRANSLATION_EVENT;

var Rosetta = (function (_EventEmitter) {
  _inherits(Rosetta, _EventEmitter);

  function Rosetta() {
    var _ref = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];

    var _ref$adapter = _ref.adapter;
    var adapter = _ref$adapter === undefined ? new _adaptersDefault2['default']() : _ref$adapter;

    _classCallCheck(this, Rosetta);

    _get(Object.getPrototypeOf(Rosetta.prototype), 'constructor', this).call(this);

    this.translator = adapter;
    this._culture = null;
    this._currency = null;
    this._languages = null;
  }

  _createClass(Rosetta, [{
    key: 't',

    // Translate.
    value: function t(key, values) {
      return this.translator.translate(key, values);
    }

    // Format number.
  }, {
    key: 'n',
    value: function n(number) {
      var options = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];

      if (typeof number !== 'number') {
        throw new Error('i18n.n should receive a number.');
      }

      return typeof Intl !== 'undefined' ? new Intl.NumberFormat(this._culture, options).format(number) : number;
    }

    // Format currency number.
  }, {
    key: 'c',
    value: function c(number) {
      var minimumFractionDigits = arguments.length <= 1 || arguments[1] === undefined ? 0 : arguments[1];

      return this.n(number, {
        style: 'currency',
        currency: this._currency,
        minimumFractionDigits: minimumFractionDigits
      });
    }
  }, {
    key: 'url',
    value: function url(urlPattern) {
      var _this = this;

      return urlPattern.split('/').map(function (token) {
        return (0, _slugify2['default'])(_this.t(token));
      }).join('/');
    }
  }, {
    key: 'addToContext',
    value: function addToContext(Component, languages) {
      return decorator.rosetta(this, languages)(Component);
    }
  }, {
    key: 'adapter',
    set: function set(adapter) {
      this.translator = adapter;
    },
    get: function get() {
      return this.translator;
    }
  }, {
    key: 'culture',
    set: function set(culture) {
      this._culture = culture;
      this.translator.locale = culture.split('-')[0];
      this.translator.translations = this._languages[culture];
    },
    get: function get() {
      return this._culture;
    }
  }, {
    key: 'currency',
    set: function set(currency) {
      this._currency = currency;
    },
    get: function get() {
      return this._currency;
    }
  }, {
    key: 'locale',
    get: function get() {
      return this.translator.locale;
    }
  }, {
    key: 'languages',
    set: function set(languages) {
      this._languages = languages;
    }
  }]);

  return Rosetta;
})(_events.EventEmitter);

exports['default'] = Rosetta;
