// @schibstedspain/sui-atom-button/lib/Button.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Button = function Button(_ref) {
  var children = _ref.children,
      href = _ref.href,
      target = _ref.target,
      disabled = _ref.disabled,
      isSubmit = _ref.isSubmit,
      isButton = _ref.isButton,
      link = _ref.link,
      Link = _ref.linkFactory,
      attrs = (0, _objectWithoutProperties3.default)(_ref, ['children', 'href', 'target', 'disabled', 'isSubmit', 'isButton', 'link', 'linkFactory']);

  if (isSubmit) attrs.type = 'submit';
  if (isButton) attrs.type = 'button';
  return link ? _react2.default.createElement(
    Link,
    (0, _extends3.default)({}, attrs, {
      href: href,
      target: target,
      rel: target === '_blank' ? 'noopener' : undefined
    }),
    children
  ) : _react2.default.createElement(
    'button',
    (0, _extends3.default)({}, attrs, { disabled: disabled }),
    children
  );
};

Button.propTypes = process.env.NODE_ENV !== "production" ? process.env.NODE_ENV !== "production" ? {
  /**
   * Content to be included in the button
   */
  children: _propTypes2.default.node,
  /**
   * HTML element: if true, render a link. Otherwise render a button
   */
  link: _propTypes2.default.bool,
  /**
   * URL to be added on the HTML link
   */
  href: _propTypes2.default.string,
  /**
   * Target to be added on the HTML link
   */
  target: _propTypes2.default.string,
  /**
   * Disable: faded with no interaction.
   */
  disabled: _propTypes2.default.bool,
  /**
   * Factory used to create navigation links
   */
  linkFactory: _propTypes2.default.func
} : {} : {};

Button.defaultProps = {
  linkFactory: function linkFactory() {
    var _ref2 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var children = _ref2.children,
        rest = (0, _objectWithoutProperties3.default)(_ref2, ['children']);
    return _react2.default.createElement(
      'a',
      rest,
      children
    );
  }
};

exports.default = Button;
